File: backend/src/db/mongodb/models/UnseenCount.ts
import mongoose, { Schema, Document, Model } from 'mongoose';
import { IUser } from '@/db/mongodb/models/User';

// Define an interface representing a UnseenCount document
export interface IUnseenCount extends Document {
    userId: mongoose.Types.ObjectId & IUser;
    mainId: mongoose.Types.ObjectId;
    count: number;
}

// Define the UnseenCount schema using the interface
const unseenCountSchema = new Schema<IUnseenCount>({
    userId: {
        type: Schema.Types.ObjectId,
        ref: 'User',
        required: true,
    },
    mainId: {
        // type: Schema.Types.ObjectId
        // ref: Chat
        // ref: Group
        type: Schema.Types.ObjectId,
        required: true,
    },
    count: {
        type: Number,
        required: true,
        default: 0,
    }

}, { timestamps: true });

// Create the UnseenCount model
const UnseenCount: Model<IUnseenCount> = mongoose.model<IUnseenCount>('UnseenCount', unseenCountSchema);

export default UnseenCount;

File: backend/consumer/models/UnseenCount.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// UnseenCount represents the unseen count document in MongoDB
type UnseenCount struct {
	UserID    primitive.ObjectID `bson:"userId" json:"userId" validate:"required"`
	MainID    primitive.ObjectID `bson:"mainId" json:"mainId" validate:"required"`
	Count     int                `bson:"count" json:"count" validate:"required,gte=0"`
	CreatedAt time.Time          `bson:"createdAt" json:"createdAt" validate:"required"`
	UpdatedAt time.Time          `bson:"updatedAt" json:"updatedAt" validate:"required"`
}
